File: src/types/events.ts
import type {
  ClusterInfo,
  ConfigMapInfo,
  CronJobInfo,
  DaemonSetInfo,
  DeploymentInfo,
  EventInfo,
  IngressInfo,
  JobInfo,
  KubernetesAction,
  KubernetesResourceKind,
  NamespaceInfo,
  NodeInfo,
  PersistentVolumeClaimInfo,
  PersistentVolumeInfo,
  PodInfo,
  PolicyInfo,
  ReplicaSetInfo,
  SecretInfo,
  ServiceInfo,
  StatefulSetInfo,
  StorageClassInfo
} from './kubernetes'

import type { UserProfileInfo } from './profile'
import type { OrbitConfig } from './settings'

export interface ComponentVersion {
  version: string
  url: string
}

export interface UpdateManifest {
  version: string
  url: string
  release_notes?: string
}

export type KubernetesResourceInfo =
  | ServiceInfo
  | IngressInfo
  | DeploymentInfo
  | PodInfo
  | StatefulSetInfo
  | DaemonSetInfo
  | ReplicaSetInfo
  | JobInfo
  | CronJobInfo
  | NodeInfo
  | NamespaceInfo
  | ConfigMapInfo
  | SecretInfo
  | EventInfo
  | PersistentVolumeInfo
  | PersistentVolumeClaimInfo
  | StorageClassInfo
  | PolicyInfo

export interface ResourceUpdateItem {
  action: KubernetesAction
  data: KubernetesResourceInfo
}

export interface OrbitEventMap {
  engineConnected: {
    status: 'ready' | 'error'
    message: string
  }
  engineTimeout: Record<string, never>
  ping: {
    message: string
  }
  pong: {
    reply: string
  }
  appSettingsUpdated: {
    settings: OrbitConfig
  }
  userProfileUpdated: {
    profile: UserProfileInfo
  }
  namespacesUpdated: {
    namespaces: NamespaceInfo[]
  }
  podsUpdated: {
    pods: PodInfo[]
  }
  deploymentsUpdated: {
    deployments: DeploymentInfo[]
  }
  statefulSetsUpdated: {
    stateful_sets: StatefulSetInfo[]
  }
  daemonSetsUpdated: {
    daemon_sets: DaemonSetInfo[]
  }
  replicaSetsUpdated: {
    replica_sets: ReplicaSetInfo[]
  }
  jobsUpdated: {
    jobs: JobInfo[]
  }
  cronJobsUpdated: {
    cron_jobs: CronJobInfo[]
  }
  clustersUpdated: {
    clusters: ClusterInfo[]
  }
  activeClusterChanged: {
    active_cluster_id: string | null
  }
  nodesUpdated: {
    nodes: NodeInfo[]
  }
  servicesUpdated: {
    services: ServiceInfo[]
  }
  ingressesUpdated: {
    ingresses: IngressInfo[]
  }
  configMapsUpdated: {
    config_maps: ConfigMapInfo[]
  }
  secretsUpdated: {
    secrets: SecretInfo[]
  }
  eventsUpdated: {
    events: EventInfo[]
  }
  persistentVolumesUpdated: {
    persistent_volumes: PersistentVolumeInfo[]
  }
  persistentVolumeClaimsUpdated: {
    persistent_volume_claims: PersistentVolumeClaimInfo[]
  }
  storageClassesUpdated: {
    storage_classes: StorageClassInfo[]
  }
  policiesUpdated: {
    policies: PolicyInfo[]
  }
  resourceBatchUpdated: {
    kind: KubernetesResourceKind | string
    updates: ResourceUpdateItem[]
  }
  podMetricsUpdated: {
    metrics: Array<{ name: string; namespace: string; cpu: string; memory: string }>
  }
  errorOccurred: {
    message: string
  }
  logLineReceived: {
    pod: string
    container: string
    line: string
  }
  logLinesChunkReceived: {
    pod: string
    container: string
    lines: string[]
  }
  updateCheckFinished: {
    has_update: boolean
    manifest: UpdateManifest
  }
  updateDownloadProgress: {
    component: string
    progress_percentage: number
  }
  updateReady: {
    component: string
  }
  commandSucceeded: {
    message: string
  }
  resourceRawData: {
    kind: string
    name: string
    data: Record<string, unknown>
  }
  portForwardStarted: {
    id: string
    namespace: string
    kind: string
    name: string
    localPort: number
    remotePort: number
  }
  portForwardStopped: {
    id: string
  }
}

/** Sentinel value for the tailLines IPC field meaning "fetch all log lines". */
export const TAIL_ALL_LINES = -1 as const

export const OrbitEvents = {
  EngineConnected: 'engineConnected',
  EngineTimeout: 'engineTimeout',
  Ping: 'ping',
  Pong: 'pong',
  AppSettingsUpdated: 'appSettingsUpdated',
  NamespacesUpdated: 'namespacesUpdated',
  PodsUpdated: 'podsUpdated',
  DeploymentsUpdated: 'deploymentsUpdated',
  StatefulSetsUpdated: 'statefulSetsUpdated',
  DaemonSetsUpdated: 'daemonSetsUpdated',
  ReplicaSetsUpdated: 'replicaSetsUpdated',
  JobsUpdated: 'jobsUpdated',
  CronJobsUpdated: 'cronJobsUpdated',
  ClustersUpdated: 'clustersUpdated',
  ActiveClusterChanged: 'activeClusterChanged',
  UserProfileUpdated: 'userProfileUpdated',
  NodesUpdated: 'nodesUpdated',
  ServicesUpdated: 'servicesUpdated',
  IngressesUpdated: 'ingressesUpdated',
  ConfigMapsUpdated: 'configMapsUpdated',
  SecretsUpdated: 'secretsUpdated',
  EventsUpdated: 'eventsUpdated',
  PersistentVolumesUpdated: 'persistentVolumesUpdated',
  PersistentVolumeClaimsUpdated: 'persistentVolumeClaimsUpdated',
  StorageClassesUpdated: 'storageClassesUpdated',
  PoliciesUpdated: 'policiesUpdated',
  ResourceBatchUpdated: 'resourceBatchUpdated',
  PodMetricsUpdated: 'podMetricsUpdated',
  ErrorOccurred: 'errorOccurred',
  LogLineReceived: 'logLineReceived',
  LogLinesChunkReceived: 'logLinesChunkReceived',
  UpdateCheckFinished: 'updateCheckFinished',
  UpdateDownloadProgress: 'updateDownloadProgress',
  UpdateReady: 'updateReady',
  CommandSucceeded: 'commandSucceeded',
  ResourceRawData: 'resourceRawData',
  PortForwardStarted: 'portForwardStarted',
  PortForwardStopped: 'portForwardStopped'
} as const

export type OrbitEventName = keyof OrbitEventMap

export type OrbitEvent<K extends OrbitEventName = OrbitEventName> = {
  event: K
  data: OrbitEventMap[K]
}

File: core/engine/src/ipc/events.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "event", content = "data")]
pub enum OrbitEvent {
    #[serde(rename = "engineConnected")]
    EngineConnected {
        status: String,
        message: String,
    },
    #[serde(rename = "ping")]
    Ping {
        message: String,
    },
    #[serde(rename = "pong")]
    Pong {
        reply: String,
    },
    #[serde(rename = "namespacesUpdated")]
    NamespacesUpdated {
        namespaces: Vec<crate::kubernetes::models::NamespaceInfo>,
    },
    #[serde(rename = "podsUpdated")]
    PodsUpdated {
        pods: Vec<crate::kubernetes::models::PodInfo>,
    },
    #[serde(rename = "deploymentsUpdated")]
    DeploymentsUpdated {
        deployments: Vec<crate::kubernetes::models::DeploymentInfo>,
    },
    #[serde(rename = "statefulSetsUpdated")]
    StatefulSetsUpdated {
        stateful_sets: Vec<crate::kubernetes::models::StatefulSetInfo>,
    },
    #[serde(rename = "daemonSetsUpdated")]
    DaemonSetsUpdated {
        daemon_sets: Vec<crate::kubernetes::models::DaemonSetInfo>,
    },
    #[serde(rename = "replicaSetsUpdated")]
    ReplicaSetsUpdated {
        replica_sets: Vec<crate::kubernetes::models::ReplicaSetInfo>,
    },
    #[serde(rename = "jobsUpdated")]
    JobsUpdated {
        jobs: Vec<crate::kubernetes::models::JobInfo>,
    },
    #[serde(rename = "cronJobsUpdated")]
    CronJobsUpdated {
        cron_jobs: Vec<crate::kubernetes::models::CronJobInfo>,
    },
    #[serde(rename = "clustersUpdated")]
    ClustersUpdated {
        clusters: Vec<crate::kubernetes::models::ClusterInfo>,
    },
    #[serde(rename = "activeClusterChanged")]
    ActiveClusterChanged {
        active_cluster_id: Option<String>,
    },
    #[serde(rename = "userProfileUpdated")]
    UserProfileUpdated {
        profile: crate::kubernetes::models::UserProfileInfo,
    },
    #[serde(rename = "nodesUpdated")]
    NodesUpdated {
        nodes: Vec<crate::kubernetes::models::NodeInfo>,
    },
    #[serde(rename = "servicesUpdated")]
    ServicesUpdated {
        services: Vec<crate::kubernetes::models::ServiceInfo>,
    },
    #[serde(rename = "ingressesUpdated")]
    IngressesUpdated {
        ingresses: Vec<crate::kubernetes::models::IngressInfo>,
    },
    #[serde(rename = "configMapsUpdated")]
    ConfigMapsUpdated {
        config_maps: Vec<crate::kubernetes::models::ConfigMapInfo>,
    },
    #[serde(rename = "secretsUpdated")]
    SecretsUpdated {
        secrets: Vec<crate::kubernetes::models::SecretInfo>,
    },
    #[serde(rename = "eventsUpdated")]
    EventsUpdated {
        events: Vec<crate::kubernetes::models::EventInfo>,
    },
    #[serde(rename = "persistentVolumesUpdated")]
    PersistentVolumesUpdated {
        persistent_volumes: Vec<crate::kubernetes::models::PersistentVolumeInfo>,
    },
    #[serde(rename = "persistentVolumeClaimsUpdated")]
    PersistentVolumeClaimsUpdated {
        persistent_volume_claims: Vec<crate::kubernetes::models::PersistentVolumeClaimInfo>,
    },
    #[serde(rename = "storageClassesUpdated")]
    StorageClassesUpdated {
        storage_classes: Vec<crate::kubernetes::models::StorageClassInfo>,
    },
    #[serde(rename = "policiesUpdated")]
    PoliciesUpdated {
        policies: Vec<crate::kubernetes::models::PolicyInfo>,
    },
    #[serde(rename = "resourceUpdated")]
    ResourceUpdated {
        kind: String,
        action: String,
        data: serde_json::Value,
    },
    #[serde(rename = "podMetricsUpdated")]
    PodMetricsUpdated {
        metrics: Vec<crate::kubernetes::models::PodMetricItem>,
    },
    #[serde(rename = "errorOccurred")]
    ErrorOccurred {
        message: String,
    },
    #[serde(rename = "logLineReceived")]
    LogLineReceived {
        pod: String,
        container: String,
        line: String,
    },
    #[serde(rename = "logLinesChunkReceived")]
    LogLinesChunkReceived {
        pod: String,
        container: String,
        lines: Vec<String>,
    },
    #[serde(rename = "updateCheckFinished")]
    UpdateCheckFinished {
        has_update: bool,
        manifest: crate::updater::UpdateManifest,
    },
    #[serde(rename = "commandSucceeded")]
    CommandSucceeded {
        message: String,
    },
    #[serde(rename = "updateDownloadProgress")]
    UpdateDownloadProgress {
        component: String,
        progress_percentage: u8,
    },
    #[serde(rename = "updateReady")]
    UpdateReady {
        component: String,
    },
    #[serde(rename = "resourceRawData")]
    ResourceRawData {
        kind: String,
        name: String,
        data: serde_json::Value,
    },
    #[serde(rename = "portForwardStarted")]
    #[serde(rename_all = "camelCase")]
    PortForwardStarted {
        id: String,
        namespace: String,
        kind: String,
        name: String,
        local_port: u16,
        remote_port: u16,
    },
    #[serde(rename = "portForwardStopped")]
    #[serde(rename_all = "camelCase")]
    PortForwardStopped {
        id: String,
    },
}

impl OrbitEvent {
    pub fn event_name(&self) -> &'static str {
        match self {
            OrbitEvent::EngineConnected { .. } => "engineConnected",
            OrbitEvent::Ping { .. } => "ping",
            OrbitEvent::Pong { .. } => "pong",
            OrbitEvent::NamespacesUpdated { .. } => "namespacesUpdated",
            OrbitEvent::PodsUpdated { .. } => "podsUpdated",
            OrbitEvent::DeploymentsUpdated { .. } => "deploymentsUpdated",
            OrbitEvent::StatefulSetsUpdated { .. } => "statefulSetsUpdated",
            OrbitEvent::DaemonSetsUpdated { .. } => "daemonSetsUpdated",
            OrbitEvent::ReplicaSetsUpdated { .. } => "replicaSetsUpdated",
            OrbitEvent::JobsUpdated { .. } => "jobsUpdated",
            OrbitEvent::CronJobsUpdated { .. } => "cronJobsUpdated",
            OrbitEvent::ClustersUpdated { .. } => "clustersUpdated",
            OrbitEvent::ActiveClusterChanged { .. } => "activeClusterChanged",
            OrbitEvent::UserProfileUpdated { .. } => "userProfileUpdated",
            OrbitEvent::NodesUpdated { .. } => "nodesUpdated",
            OrbitEvent::ServicesUpdated { .. } => "servicesUpdated",
            OrbitEvent::IngressesUpdated { .. } => "ingressesUpdated",
            OrbitEvent::ConfigMapsUpdated { .. } => "configMapsUpdated",
            OrbitEvent::SecretsUpdated { .. } => "secretsUpdated",
            OrbitEvent::EventsUpdated { .. } => "eventsUpdated",
            OrbitEvent::PersistentVolumesUpdated { .. } => "persistentVolumesUpdated",
            OrbitEvent::PersistentVolumeClaimsUpdated { .. } => "persistentVolumeClaimsUpdated",
            OrbitEvent::StorageClassesUpdated { .. } => "storageClassesUpdated",
            OrbitEvent::PoliciesUpdated { .. } => "policiesUpdated",
            OrbitEvent::ResourceUpdated { .. } => "resourceUpdated",
            OrbitEvent::PodMetricsUpdated { .. } => "podMetricsUpdated",
            OrbitEvent::ErrorOccurred { .. } => "errorOccurred",
            OrbitEvent::LogLineReceived { .. } => "logLineReceived",
            OrbitEvent::LogLinesChunkReceived { .. } => "logLinesChunkReceived",
            OrbitEvent::UpdateCheckFinished { .. } => "updateCheckFinished",
            OrbitEvent::CommandSucceeded { .. } => "commandSucceeded",
            OrbitEvent::UpdateDownloadProgress { .. } => "updateDownloadProgress",
            OrbitEvent::UpdateReady { .. } => "updateReady",
            OrbitEvent::ResourceRawData { .. } => "resourceRawData",
            OrbitEvent::PortForwardStarted { .. } => "portForwardStarted",
            OrbitEvent::PortForwardStopped { .. } => "portForwardStopped",
        }
    }
}
